content/en/docs/architecture/ci-operator-internals/testing.md

```diff
@@ -0,0 +1,425 @@
+{{< / highlight >}}
+
+The CI build clusters are heavily used, so it is best to provide resource
+definitions to guide the scheduling of test processes, which is done using the
+(required) `resources` field.  It is a mapping field whose key is a name and
+whose value is a definition of resources required by the test `Pod`, as
+[specified by Kubernetes][kubernetes_resources].  A unique entry for the test
+can be added under the `test` key (the name of the test), or a generic one can
+be added using the special `*` key, which defines a default.
+
+{{< highlight yaml >}}
+# test.yaml
+tests:
+  # as before
+resources:
+  "*":
+    requests:
+      cpu: 100m
+      memory: 8M # ought to be enough for anybody
+{{< / highlight >}}
+
+{{< highlight bash >}}
+$ ci-operator --config test.yaml
+INFO[2022-03-09T18:15:32Z] unset version 0
+ERRO[2022-03-09T18:15:32Z] Failed to load arguments.                     error=tests[test].from: unknown image "quay.io/fedora/fedora"
+{{< / highlight >}}
+
+The image declaration (`from`) for the container test will not work as defined.
+This is because, in order to determine the state of all the inputs listed in the
+configuration file, `ci-operator` requires that all resources be local to the
+cluster.  For this reason, direct references to external images (e.g. from Quay)
+such as the one we are attempting are not allowed.  The process for [configuring
+image inputs][image_inputs] and [using external images][external_images] is
+detailed elsewhere.  For now, we can use an image that is known to already exist
+in our clusters.  Declaring cluster images which will be used in the
+configuration is done using `base_images`:
+
+{{< highlight yaml >}}
+base_images:
+  fedora:
+    namespace: ci
+    name: fedora
+    tag: latest
+tests:
+- as: test
+  commands: echo test
+  container:
+    from: fedora
+    clone: false
+resources:
+  # as before
+{{< / highlight >}}
+
+`clone: false` was added since by default tests which use one of the
+`base_images` have the source code of the repository automatically cloned.  Our
+configuration file is finally valid and ready to guide the execution of the test
+it defines, but a few topics need to be covered before we can see the result.
+
+## Test namespaces
+
+Even though `ci-operator` is being executed directly on our machine, that is not
+where most of the actual work is done.  Image builds, test executions, etc. must
+happen inside an Openshift cluster, and `ci-operator` expects to be executed in
+an environment where it has access to one.  This is why this final error message
+mentions "in-cluster configuration" and missing Kuberentes host/port
+information.  As mentioned before, the default environment is assumed to be a
+Prow job, which is executed inside a Kubernetes `Pod`.  There, those environment
+variables are provided automatically by Kubernetes (see the [service
+documentation][services_networking]).
```

Comment:
Hmm, that document must have changed since I wrote this.  This is meant to show the expected environment in actual jobs, so changed to one that explains those variables.